rfcs/QueryLevelNullability.md

```diff
@@ -0,0 +1,179 @@
+
+This RFC proposes creating a syntactical construct for client developers to 
+express "nullability" in their queries.
+
+## Definitions
+
+Nullability: A concept that exists across many programming languages (eg [Swift](https://developer.apple.com/documentation/swift/optional), [Kotlin](https://kotlinlang.org/docs/null-safety.html#nullable-types-and-non-null-types), [SQL](https://www.w3schools.com/sql/sql_notnull.asp))
+that is used to express when users can be certain that a value can or can never be `null` 
+(or the language equivalent). Nullability language constructs (eg `?` in Swift/Kotlin)
+have become popular due to their ability to solve ergonomic problems in languages
+such as those surrounding `NullPointerException` in Java.
+
+Codegen: Short for Code Generation, tools that exist for working with GraphQL which take queries and schemas as
+input and output code in a language of choice that represents data that will result from those 
+queries. Codegen tools exist for many platforms. As an example, [here's some info about Apollo's offerings.](https://github.com/apollographql/apollo-tooling#code-generation)
+
+## 📜 Problem Statement
+
+GraphQL is a "nullable by default" language meaning that all properties are allowed to be `null`.
+This is in contrast to the two modern languages used on mobile clients, Swift and Kotlin,
+which are both non-null by default. In Swift and Kotlin, unless developers otherwise
+specify it, properties cannot be `null`.
+
+This mismatch creates some dissonance for developers who are currently forced into dealing the
+problems that commonly surround nullablility in codebases that otherwise do not need to deal with
+those problems. This makes large numbers of nulls tedious and time-consuming to handle.
+
+It is often unclear how to handle partial results.
+- What elements are essential to having an adequate user experience? How many elements can fail before you 
+  should just show an error message instead?
+- When any combination of elements can fail, it is very hard to anticipate all edge cases and how the app 
+  might look and work, including transitions to other screens.
+
+In Yelp's GraphQL schema, almost all object fields are nullable except for those with ID type. 
+This adheres to what seems to be the [official best practice](https://graphql.org/learn/best-practices/#nullability).
+
+This poses a problem for the mobile clients that use Apollo's codegen feature. The codegen provides 
+Swift/Kotlin types to represent queries used in the app. Nullable fields in the schema are represented
+as nullable properties on the resulting type, represented by `?` following the type name:
+```graphql
+query GetBusinessName($encid: String!) {
+  business(encid: $encid) {
+    name
+  }
+}
+```
+```swift
+struct GetBusinessNameQuery.Data.Business {
+  let name: String?
+}
+```
+In many cases, the client should error if the business `name` is null. If codegen were out of the picture,
```

Comment:
It took me several rereads to understand the described use case. Is this a correct understanding? 

While the schema can have the nullable fields for valid reasons (such as federation), the client wants to decide if it accepts a null value for the result to simplify the client-side logic.

Reply:
Yeah that's right! I took your wording since it's more generally applicable than what we had.